fix(ui): flush ext_cmdline events before doing cmdpreview

luukvbaal opened the pull request:
Problem:  Unable to update the screen for external cmdline during cmdpreview.
Solution: Flush the cmdline UI before cmdpreview state.

Resolve https://github.com/neovim/neovim/issues/20463

bfredl commented on src/nvim/ex_getln.c:
```diff
@@ -3754,7 +3759,7 @@ void cmdline_paste_str(const char *s, bool literally)
 void redrawcmdline(void)
 {
-  if (cmd_silent) {
+  if (ui_has(kUICmdline) || cmd_silent) {
```
> This breaks a few tested UI events.

then `redrawcmdline()` is being called as a result of the cmdline _state_ changed as well. might need a bool flag to distinguish these two if we want to prune unnecessary events.

luukvbaal replied:
Not sure what you mean, the broken tests specifically tested for re-emitted events after `:mode`.

bfredl commented on src/nvim/ex_getln.c:
```diff
@@ -3458,6 +3434,9 @@ void cmdline_ui_flush(void)
     if (line->level == level) {
       if (line->redraw_state == kCmdRedrawAll) {
         ui_ext_cmdline_show(line);
+        if (ui_did_lua_cb) {
+          update_screen();
+        }
```
I don't like thisl. it would be better (and generalizable to other usecases) to add whatever `vim.redraw_...()` API to lua so the callback can add redraws as needed.

luukvbaal replied:
OK yeah sure, if we are OK with exposing `update_screen()` I'll look into a `vim.redraw`.

justinmk reviewed (approved):
Nice code reduction. @bfredl any objections to this?

luukvbaal commented:
Rebased and reduced this PR to just flush before doing cmdpreview, which I think still addresses the linked issue. (The other changes here in stale reviews may have been unnecessary.)

Fully resolving it requires removing the intentional redraw restriction that is in place by making the cmdpreview screen state non-ephemeral (planned work by @theofabilous / @famiu: #28510).

Once that is done this flush may be removed but I think this may still be nice to have as an intermediate fix that is sufficient to allow updating the screen with new cmdline state, which I think was the premise of the linked issue.

luukvbaal commented:
> that is sufficient to allow updating the screen with new cmdline state

No longer true after rebasing on current master.

luukvbaal commented:
> Rebasing on master however the drawn cmdline lags one character behind as you say.

I'm not sure what happened when I tried rebasing this earlier but retesting now it still seems to work as (I) expected with the simple repro from the the added test, also interactively.

> With your PR, I get a segfault after entering `:%s` if I do a create/update windows in the `ext_cmdline` callback(from https://github.com/neovim/neovim/issues/28772#issuecomment-2115393628).

I couldn't reproduce a segfault but did hit an `abort()` in `cmdpreview_restore_state()` when using noice with this PR. Moving the flush up a little prevents that, could you try again and provide repro steps if this still doesn't work for you @folke?

folke commented:
Yes, it works great now!!

And I no longer see any segfaults, but that might be because I also had to add in some code (not for this PR) that prevents ui updates during textlock (which I unfortunately need to check with ffi). textlock is what was triggering all the new E565 errors I kept seeing.

luukvbaal committed:
fix(ui): flush ext_cmdline events before doing cmdpreview

Problem:  Unable to update the screen for external cmdline during cmdpreview.
Solution: Flush the cmdline UI before cmdpreview state.

folke commented:
Would be great to have this merged. Is there anything blocking this right now?

luukvbaal commented:
I don't think so. This will become redundant in case #28856 pans out. But even if it does, this seems safer to backport to 0.10, if we want to do that.
